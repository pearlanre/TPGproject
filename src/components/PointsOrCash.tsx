import React, { useEffect, useState } from 'react';
import Select from 'react-select';
import CurrencyInput from 'react-currency-input';
import './PointsOrCash.css';

// eslint-disable-next-line @typescript-eslint/no-empty-interface
interface Props {}

export const PointsOrCash: React.FunctionComponent<Props> = () => {
  const [data, setData] = useState<any[]>([]);
  const getData = () => {
    fetch('data.json', {
      headers: {
        'Content-Type': 'application/json',
        Accept: 'application/json',
      },
    })
      .then(function (response) {
        return response.json();
      })
      .then(function (myJson) {
        setData(myJson);
      });
  };
  useEffect(() => {
    getData();
  }, []);
  const [loyaltyProgram, setLoyaltyProgram] = useState({
    name: '',
    tpg_valuation: 0,
  });
  const [ticketCost, setTicketCost] = useState('');
  const [points, setPoints] = useState(0);
  const [submitted, setSubmitted] = useState(false);
  const [pointsPrice, setPointsPrice] = useState('');

  const handleSubmit = (e: any) => {
    e.preventDefault();

    const total =
      parseFloat(String(loyaltyProgram.tpg_valuation)) *
      parseInt(String(points), 10);
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    setPointsPrice(total);
    setSubmitted(true);
  };
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  return (
    <div>
      <form className="register-form" onSubmit={handleSubmit}>
        <p>Select rewards Program</p>
        {data && data.length > 0 && (
          <Select
            options={data}
            getOptionLabel={(option) => option.name}
            getOptionValue={(option) => option.id}
            defaultValue={loyaltyProgram}
            onChange={setLoyaltyProgram}
          />
        )}

        {submitted && !loyaltyProgram && (
          <span id="loyalty-error">Please select a Loyalty program</span>
        )}
        <p>Enter ticker price</p>
        <CurrencyInput
          className="form-field"
          value={ticketCost}
          onChangeEvent={(e) => setTicketCost(e.target.value)}
        />
        {submitted && !ticketCost && (
          <span id="loyalty-error">Please Enter the Ticket Cost</span>
        )}
        <p>How many points do you have ?</p>
        <CurrencyInput
          className="form-field"
          value={points}
          onChangeEvent={(e) => setPoints(e.target.value)}
        />
        {submitted && (
          <span id="loyalty-error">
            The program valuations for {loyaltyProgram.name} has tpg value{' '}
            {loyaltyProgram.tpg_valuation} results to {pointsPrice}
          </span>
        )}
        <button className="form-field" type="submit">
          Calculate
        </button>
        {pointsPrice > ticketCost ? (
          <p>You should use Cash for this ticket</p>
        ) : (
          <p>You should use points for this ticket</p>
        )}
      </form>
    </div>
  );
};

export default PointsOrCash;
