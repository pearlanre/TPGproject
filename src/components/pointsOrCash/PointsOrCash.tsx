import React, { useEffect, useState } from 'react';
import Select from 'react-select';
import CurrencyInput from 'react-currency-input';
import './PointsOrCash.css';

// eslint-disable-next-line @typescript-eslint/no-empty-interface
interface Props {}

export const PointsOrCash: React.FunctionComponent<Props> = () => {
  const [data, setData] = useState<any[]>([]);
  const getData = () => {
    fetch('data.json', {
      headers: {
        'Content-Type': 'application/json',
        Accept: 'application/json',
      },
    })
      .then(function (response) {
        return response.json();
      })
      .then(function (myJson) {
        const filtered = myJson.filter((val) => val.tpg_valuation !== '');
        setData(filtered);
      });
  };
  useEffect(() => {
    getData();
  }, []);
  const [loyaltyProgram, setLoyaltyProgram] = useState({
    name: '',
    tpg_valuation: 0,
  });
  const [ticketCost, setTicketCost] = useState(0);
  const [submitted, setSubmitted] = useState(false);
  const [ticketFees, setTicketFees] = useState(5.95);
  const [usePoints, setUsePoints] = useState(false);
  const [pointsNeeded, setPointsNeeded] = useState(0);

  const handleSubmit = (e: any) => {
    e.preventDefault();
    setSubmitted(true);
    if (loyaltyProgram.tpg_valuation > 1 && ticketFees < ticketCost) {
      setUsePoints(true);
    } else {
      setUsePoints(false);
    }
    const points = Math.round(ticketCost / loyaltyProgram.tpg_valuation);
    setPointsNeeded(points);
  };
  const handleLoyaltyChange = (option) => {
    setLoyaltyProgram(option);
    setTicketFees(5.95);
    setTicketCost(0);
    setSubmitted(false);
  };
  const showRecommendation =
    submitted && !Number.isNaN(pointsNeeded) && Number.isFinite(pointsNeeded);
  return (
    <div>
      <form className="register-form" onSubmit={handleSubmit}>
        <p>Select A Rewards Program</p>
        {data && data.length > 0 && (
          <Select
            options={data}
            getOptionLabel={(option) => option.name}
            getOptionValue={(option) => option.id}
            defaultValue={loyaltyProgram}
            onChange={handleLoyaltyChange}
          />
        )}

        {submitted && !loyaltyProgram.name && (
          <span id="loyalty-error">Please select a Loyalty program</span>
        )}
        <p>Enter Ticket Price</p>
        <CurrencyInput
          className="form-field"
          value={ticketCost}
          onChangeEvent={(e) => setTicketCost(e.target.value)}
        />
        {submitted && !ticketCost && (
          <span id="loyalty-error">Please Enter the Ticket Cost</span>
        )}
        <p>Enter Ticket Fees</p>
        <CurrencyInput
          className="form-field"
          value={ticketFees}
          onChangeEvent={(e) => setTicketFees(e.target.value)}
        />
        <i>
          * Ticket Fees are set to 5.95 by default. Please update with ticket
          Fees.
        </i>
        <button className="form-button" type="submit">
          Calculate
        </button>
        {showRecommendation && (
          <p>
            You need {pointsNeeded} points plus ${ticketFees} for this ticket.
          </p>
        )}

        {showRecommendation &&
          (usePoints ? (
            <p>TPG recommends using points for this ticket</p>
          ) : (
            <p>TPG recommends using Cash for this ticket</p>
          ))}
      </form>
    </div>
  );
};

export default PointsOrCash;
