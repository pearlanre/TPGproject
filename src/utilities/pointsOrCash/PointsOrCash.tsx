import React, { useEffect, useState } from 'react';
import Select from 'react-select';
import CurrencyInput from 'react-currency-input';
import './PointsOrCash.css';
import { ILoyalty } from '../../interfaces/ILoyalty';

// eslint-disable-next-line @typescript-eslint/no-empty-interface
interface Props {
  data: ILoyalty;
  program: string;
}
export const PointsOrCash: React.FunctionComponent<Props> = (props: Props) => {
  const [loyaltyProgram, setLoyaltyProgram] = useState({
    name: '',
    tpg_valuation: 0,
  });
  const [programCost, setProgramCost] = useState(0);
  const [submitted, setSubmitted] = useState(false);
  const [programFees, setProgramFees] = useState(5.95);
  const [usePoints, setUsePoints] = useState(false);
  const [pointsNeeded, setPointsNeeded] = useState(0);

  const handleSubmit = (e: any) => {
    e.preventDefault();
    setSubmitted(true);
    if (loyaltyProgram.tpg_valuation > 1 && programFees < programCost) {
      setUsePoints(true);
    } else {
      setUsePoints(false);
    }
    const points = Math.round(programCost / loyaltyProgram.tpg_valuation);
    setPointsNeeded(points);
  };
  const handleLoyaltyChange = (option) => {
    setLoyaltyProgram(option);
    setProgramFees(5.95);
    setProgramCost(0);
    setSubmitted(false);
  };
  const showRecommendation =
    submitted && !Number.isNaN(pointsNeeded) && Number.isFinite(pointsNeeded);
  return (
    <div>
      <form className="cash-form" onSubmit={handleSubmit}>
        {props.data && (
          <Select
            options={props.data}
            getOptionLabel={(option) => option.name}
            getOptionValue={(option) => option.id}
            defaultValue={loyaltyProgram}
            onChange={handleLoyaltyChange}
            className="form-select"
          />
        )}
        {submitted && !loyaltyProgram.name && (
          <span id="loyalty-error">Please select a Loyalty program</span>
        )}
        <p>Enter {props.program} Price</p>
        <CurrencyInput
          className="form-field"
          value={programCost}
          onChangeEvent={(e) => setProgramCost(e.target.value)}
        />
        {submitted && !loyaltyProgram.name && (
          <span id="loyalty-error">Please Enter the {props.program} Cost</span>
        )}
        <p>Enter {props.program} Fees</p>
        <CurrencyInput
          className="form-field"
          value={programFees}
          onChangeEvent={(e) => setProgramFees(e.target.value)}
        />
        <i>
          * Program Fees are set to 5.95 by default. Please update with{' '}
          {props.program} Fees.
        </i>
        <button className="form-button" type="submit">
          Calculate
        </button>
        {showRecommendation && (
          <p>
            You need {pointsNeeded} points plus ${programFees} for this{' '}
            {props.program}.
          </p>
        )}

        {showRecommendation &&
          (usePoints ? (
            <p>
              TPG recommends using <b>Points</b> for this {props.program}.
            </p>
          ) : (
            <p>
              TPG recommends using <b>Cash</b> for this {props.program}.
            </p>
          ))}
      </form>
    </div>
  );
};

export default PointsOrCash;
