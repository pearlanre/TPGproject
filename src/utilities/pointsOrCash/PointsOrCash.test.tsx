import React from 'react';
import { shallow, mount } from 'enzyme';
import CurrencyInput from 'react-currency-input';
import { PointsOrCash } from './PointsOrCash';

describe('PointsOrCash', () => {
  let props = {
    data: {
      name: 'Private airline',
      tpg_valuation: 2,
      type: '',
    },
    program: '',
  };

  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<PointsOrCash {...props} />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
  it('should render the PointsOrCash Component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('should render PointOrCash in preview mode', () => {
    expect(wrapper).toMatchSnapshot();
    wrapper.find('form').simulate('submit', {
      preventDefault: () => {},
    });
    expect(wrapper).toMatchSnapshot();
  });
  describe('test functionality for Tickets', () => {
    props = {
      data: {
        name: 'Private airline',
        tpg_valuation: 2,
        type: 'airline',
      },
      program: 'Tickets',
    };
    wrapper = shallow(<PointsOrCash {...props} />);
    it('renders ticket price label', () => {
      const ticketLabel = 'Enter Hotels Price';
      expect(wrapper.find('p').at(0).text()).toEqual(ticketLabel);
    });
    it('renders form component and other components', () => {
      expect(wrapper.find('form')).toHaveLength(1);
      expect(wrapper.find('p')).toHaveLength(2);
      expect(wrapper.find('button')).toHaveLength(1);
    });
    it('renders currency Input', () => {
      const onChange = jest.fn();
      const currency = wrapper.find(CurrencyInput);
      expect(currency.exists()).toEqual(true);
      expect(currency).toHaveLength(2);
      expect(wrapper.find('i')).toHaveLength(1);
    });

    it('should submit on when from is submitted', () => {
      const setSubmitted = jest.fn();
      const handleSubmit = jest.spyOn(React, 'useState');
      // @ts-ignore
      handleSubmit.mockImplementation((submitted) => [submitted, setSubmitted]);
      const form = wrapper.find('form');
      form.simulate('submit', {
        preventDefault: () => {},
      });
      expect(setSubmitted).toBeTruthy();
    });
  });
  describe('test functionality for Hotels', () => {
    props = {
      data: {
        name: 'Private airline',
        tpg_valuation: 2,
        type: 'airline',
      },
      program: 'Hotels',
    };
    wrapper = shallow(<PointsOrCash {...props} />);
    it('renders Hotels price label', () => {
      const hotelLabel = 'Enter Hotels Price';
      expect(wrapper.find('p').at(0).text()).toEqual(hotelLabel);
    });
    it('renders form component and other components', () => {
      expect(wrapper.find('form')).toHaveLength(1);
      expect(wrapper.find('p')).toHaveLength(2);
      expect(wrapper.find('button')).toHaveLength(1);
    });
    it('renders currency Input', () => {
      const onChange = jest.fn();
      const currency = wrapper.find(CurrencyInput);
      expect(currency.exists()).toEqual(true);
      expect(currency).toHaveLength(2);
      expect(wrapper.find('i')).toHaveLength(1);
    });

    it('should submit on when from is submitted', () => {
      const setSubmitted = jest.fn();
      const handleSubmit = jest.spyOn(React, 'useState');
      // @ts-ignore
      handleSubmit.mockImplementation((submitted) => [submitted, setSubmitted]);
      const form = wrapper.find('form');
      form.simulate('submit', {
        preventDefault: () => {},
      });
      expect(setSubmitted).toBeTruthy();
    });
  });
});
