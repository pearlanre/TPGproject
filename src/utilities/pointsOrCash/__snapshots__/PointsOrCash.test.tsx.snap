// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PointsOrCash should render PointOrCash in preview mode 1`] = `
<div>
  <form
    className="cash-form"
    onSubmit={[Function]}
  >
    <StateManager
      className="form-select"
      defaultInputValue=""
      defaultMenuIsOpen={false}
      defaultValue={
        Object {
          "name": "",
          "tpg_valuation": 0,
        }
      }
      getOptionLabel={[Function]}
      getOptionValue={[Function]}
      onChange={[Function]}
      options={
        Object {
          "name": "Private airline",
          "tpg_valuation": 2,
          "type": "airline",
        }
      }
    />
    <p>
      Enter 
      Hotels
       Price
    </p>
    <CurrencyInput
      allowNegative={false}
      autoFocus={false}
      className="form-field"
      decimalSeparator="."
      inputType="text"
      onChange={[Function]}
      onChangeEvent={[Function]}
      precision="2"
      prefix=""
      selectAllOnFocus={false}
      suffix=""
      thousandSeparator=","
      value={0}
    />
    <p>
      Enter 
      Hotels
       Fees
    </p>
    <CurrencyInput
      allowNegative={false}
      autoFocus={false}
      className="form-field"
      decimalSeparator="."
      inputType="text"
      onChange={[Function]}
      onChangeEvent={[Function]}
      precision="2"
      prefix=""
      selectAllOnFocus={false}
      suffix=""
      thousandSeparator=","
      value={5.95}
    />
    <i>
      * Program Fees are set to 5.95 by default. Please update with
       
      Hotels
       Fees.
    </i>
    <button
      className="form-button"
      type="submit"
    >
      Calculate
    </button>
  </form>
</div>
`;

exports[`PointsOrCash should render PointOrCash in preview mode 2`] = `
<div>
  <form
    className="cash-form"
    onSubmit={[Function]}
  >
    <StateManager
      className="form-select"
      defaultInputValue=""
      defaultMenuIsOpen={false}
      defaultValue={
        Object {
          "name": "",
          "tpg_valuation": 0,
        }
      }
      getOptionLabel={[Function]}
      getOptionValue={[Function]}
      onChange={[Function]}
      options={
        Object {
          "name": "Private airline",
          "tpg_valuation": 2,
          "type": "airline",
        }
      }
    />
    <span
      id="loyalty-error"
    >
      Please select a Loyalty program
    </span>
    <p>
      Enter 
      Hotels
       Price
    </p>
    <CurrencyInput
      allowNegative={false}
      autoFocus={false}
      className="form-field"
      decimalSeparator="."
      inputType="text"
      onChange={[Function]}
      onChangeEvent={[Function]}
      precision="2"
      prefix=""
      selectAllOnFocus={false}
      suffix=""
      thousandSeparator=","
      value={0}
    />
    <span
      id="loyalty-error"
    >
      Please Enter the 
      Hotels
       Cost
    </span>
    <p>
      Enter 
      Hotels
       Fees
    </p>
    <CurrencyInput
      allowNegative={false}
      autoFocus={false}
      className="form-field"
      decimalSeparator="."
      inputType="text"
      onChange={[Function]}
      onChangeEvent={[Function]}
      precision="2"
      prefix=""
      selectAllOnFocus={false}
      suffix=""
      thousandSeparator=","
      value={5.95}
    />
    <i>
      * Program Fees are set to 5.95 by default. Please update with
       
      Hotels
       Fees.
    </i>
    <button
      className="form-button"
      type="submit"
    >
      Calculate
    </button>
  </form>
</div>
`;

exports[`PointsOrCash should render the PointsOrCash Component correctly 1`] = `
<div>
  <form
    className="cash-form"
    onSubmit={[Function]}
  >
    <StateManager
      className="form-select"
      defaultInputValue=""
      defaultMenuIsOpen={false}
      defaultValue={
        Object {
          "name": "",
          "tpg_valuation": 0,
        }
      }
      getOptionLabel={[Function]}
      getOptionValue={[Function]}
      onChange={[Function]}
      options={
        Object {
          "name": "Private airline",
          "tpg_valuation": 2,
          "type": "airline",
        }
      }
    />
    <p>
      Enter 
      Hotels
       Price
    </p>
    <CurrencyInput
      allowNegative={false}
      autoFocus={false}
      className="form-field"
      decimalSeparator="."
      inputType="text"
      onChange={[Function]}
      onChangeEvent={[Function]}
      precision="2"
      prefix=""
      selectAllOnFocus={false}
      suffix=""
      thousandSeparator=","
      value={0}
    />
    <p>
      Enter 
      Hotels
       Fees
    </p>
    <CurrencyInput
      allowNegative={false}
      autoFocus={false}
      className="form-field"
      decimalSeparator="."
      inputType="text"
      onChange={[Function]}
      onChangeEvent={[Function]}
      precision="2"
      prefix=""
      selectAllOnFocus={false}
      suffix=""
      thousandSeparator=","
      value={5.95}
    />
    <i>
      * Program Fees are set to 5.95 by default. Please update with
       
      Hotels
       Fees.
    </i>
    <button
      className="form-button"
      type="submit"
    >
      Calculate
    </button>
  </form>
</div>
`;
